river/rules/base.py

```diff
@@ -0,0 +1,275 @@
+import abc
+import copy
+import math
+import numbers
+import typing
+
+from river import base, tree
+
+
+class Literal(base.Base):
+    __slots__ = "on", "at", "neg"
+
+    def __init__(self, on, at, neg=False):
+        self.on = on
+        self.at = at
+        self.neg = neg
+
+    @abc.abstractmethod
+    def __call__(self, x):
+        pass
+
+    @abc.abstractmethod
+    def describe(self):
+        pass
+
+
+class NumericLiteral(Literal):
+    def __init__(self, on, at, neg):
+        super().__init__(on, at, neg)
+
+    def __call__(self, x):
+        if self.on in x:
+            if not self.neg:
+                return x[self.on] <= self.at
+            else:
+                return x[self.on] > self.at
+
+        return False
+
+    def describe(self):
+        if not self.neg:
+            return f"{self.on} ≤ {self.at}"
+        else:
+            return f"{self.on} > {self.at}"
+
+
+class NominalLiteral(Literal):
+    def __init__(self, on, at, neg):
+        super().__init__(on, at, neg)
+
+    def __call__(self, x):
+        if self.on in x:
+            if not self.neg:
+                return x[self.on] == self.at
+            else:
+                return x[self.on] != self.at
+
+        return False
+
+    def describe(self):
+        if not self.neg:
+            return f"{self.on} = {self.at}"
+        else:
+            return f"{self.on} ≠ {self.at}"
+
+
+class HoeffdingRule(base.Estimator, metaclass=abc.ABCMeta):
```

Comment:
Please provide a short description.

Reply:
I added some additional descriptions in commit [2745f34](https://github.com/online-ml/river/pull/594/commits/2745f3495b9dc50ecb30f62be4230fd05f17499d).